File: Python/programmers/etc/level2/가장_큰_정사각형_찾기.py
def solution(board):
    answer = 0
    n, m = len(board), len(board[0])
    dp = [[0]*(m+1) for _ in range(n+1)]
    for i in range(1, n+1):
        for j in range(1, m+1):
            if board[i-1][j-1] == 1:
                dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])+1
                answer = max(answer, dp[i][j])
    return answer**2


board = [[0,1,1,1],[1,1,1,1],[1,1,1,1],[0,0,1,0]]
print(solution(board)) # 9

board = [[0,0,1,1],[1,1,1,1]]
print(solution(board)) # 4

File: Java/programmers/etc/level2/가장큰정사각형찾기.java
package CodeTest.Java.programmers.etc.level2;

public class 가장큰정사각형찾기 {
    public int solution(int[][] board) {
        int answer = 0;
        int n = board.length, m = board[0].length;
        int[][] dp = new int[n+1][m+1];

        for (int i=1; i<n+1; i++) {
            for (int j=1; j<m+1; j++) {
                if (board[i-1][j-1] == 1) {
                    dp[i][j] = Math.min(dp[i-1][j], Math.min(dp[i][j-1], dp[i-1][j-1]))+1;
                    answer = Math.max(answer, dp[i][j]);
                }
            }
        }

        return answer*answer;
    }
}
